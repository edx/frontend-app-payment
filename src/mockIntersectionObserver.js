global.IntersectionObserver = class IntersectionObserver {
  constructor(callback) {
    this.callback = callback;
    this.observedElements = new Set();
  }

  observe(element) {
    this.callback([{ isIntersecting: true }]);
    this.observedElements.add(element);
  }

  unobserve(element) {
    this.observedElements.delete(element);
  }

  disconnect() {
    this.observedElements.clear();
  }
};
