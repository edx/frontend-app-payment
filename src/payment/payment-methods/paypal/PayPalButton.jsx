import React, { useEffect, useRef } from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';
import { trackElementIntersection } from '../../data/actions';
import { ElementType, IS_FULLY_SHOWN_THRESHOLD_OR_MARGIN } from '../../../cohesion/constants';

import PayPalLogo from './assets/paypal-logo.png';
import messages from './PayPalButton.messages';

const PayPalButton = ({ intl, isProcessing, ...props }) => {
  const buttonRef = useRef(null);
  const dispatch = useDispatch();

  // RV event tracking for PayPal Button
  useEffect(() => {
    const observerCallback = (entries) => {
      entries.forEach(entry => {
        if (entry.isIntersecting) {
          const elementId = entry.target?.id;
          const tagularElement = {
            elementType: ElementType.Button,
            position: String(elementId || ''),
            name: 'paypal',
            text: 'PayPal',
          };
          dispatch(trackElementIntersection(tagularElement));
        }
      });
    };

    const observer = new IntersectionObserver(observerCallback, {
      threshold: IS_FULLY_SHOWN_THRESHOLD_OR_MARGIN,
    });

    const currentElement = buttonRef.current;

    if (currentElement) {
      observer.observe(currentElement);
    }

    return () => {
      if (currentElement) {
        observer.unobserve(currentElement);
      }
      observer.disconnect();
    };
  }, [dispatch]);

  return (
    <button type="button" {...props} ref={buttonRef}>
      { isProcessing ? <span className="button-spinner-icon text-primary mr-2" /> : null }
      <img
        src={PayPalLogo}
        alt={intl.formatMessage(messages['payment.type.paypal'])}
      />
    </button>
  );
};

PayPalButton.propTypes = {
  intl: intlShape.isRequired,
  isProcessing: PropTypes.bool,
};

PayPalButton.defaultProps = {
  isProcessing: false,
};

export default injectIntl(PayPalButton);
