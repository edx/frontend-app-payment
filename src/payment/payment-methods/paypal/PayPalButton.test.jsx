import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { render } from '@testing-library/react';
import { IntlProvider } from '@edx/frontend-platform/i18n';

import '../../../mockIntersectionObserver';

import PayPalButton from './PayPalButton';

const mockStore = configureMockStore();

describe('PayPalButton', () => {
  let store;
  let state;

  beforeEach(() => {
    state = {
      userAccount: { email: 'person@example.com' },
      payment: {
        basket: {
          loaded: false,
          loading: false,
          products: [],
        },
      },
      i18n: {
        locale: 'en',
      },
    };

    store = mockStore(state);
  });

  it('should render the button by default', () => {
    const component = (
      <IntlProvider locale="en">
        <Provider store={store}>
          <PayPalButton />
        </Provider>
      </IntlProvider>
    );
    const { container: tree } = render(component);
    expect(tree).toMatchSnapshot();
  });
  it('should render the button with a spinner when processing', () => {
    const component = (
      <IntlProvider locale="en">
        <Provider store={store}>
          <PayPalButton isProcessing />
        </Provider>
      </IntlProvider>
    );
    const { container: tree } = render(component);
    expect(tree).toMatchSnapshot();
  });
});
