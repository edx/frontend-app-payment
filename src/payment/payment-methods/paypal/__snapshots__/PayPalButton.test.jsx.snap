// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PayPalButton should render the button by default 1`] = `
<div>
  <button
    type="button"
  >
    <img
      alt="PayPal"
      src="test-file-stub"
    />
  </button>
</div>
`;

exports[`PayPalButton should render the button with a spinner when processing 1`] = `
<div>
  <button
    type="button"
  >
    <span
      class="button-spinner-icon text-primary mr-2"
    />
    <img
      alt="PayPal"
      src="test-file-stub"
    />
  </button>
</div>
`;
